Added google call creds/per_rpc interop tests

kpayson64 opened the pull request:
This is a breakout of the per_rpc_creds from #6155.
Adds a wrapper for google creds, as well as implementing the per_rpc interop tests in issue #4389

jtattermusch commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,69 @@
+from grpc.beta import interfaces
+
+class GoogleCallCredentials(interfaces.GRPCAuthMetadataPlugin):
+  """ Metadata wrapper for googleCredentials from the oauth2client library
```
nit: I think the comment should be """Some comment with no spaces around""", if it fits on one line. Also below.

jtattermusch commented on src/python/grpcio/grpc/beta/implementations.py:
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
   return CallCredentials(
       _low.call_credentials_metadata_plugin(metadata_plugin, name))
 
+def google_call_credentials(credentials):
```
I think the naming "google_call_credentials" and "access_token_call_credential" is good, let's get a second opinion from @nathanielmanistaatgoogle  or @soltanmm.  If everything else looks good, I think it's fine to submit this change and revisit the naming if necessary.

jtattermusch commented:
LGTM after fixing the comment format I mentioned.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,72 @@
+# Copyright 2015, Google Inc.
```
Header copied from last year.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,72 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from concurrent.futures import ThreadPoolExecutor
```
["Use imports for packages and modules only."](https://google.github.io/styleguide/pyguide.html?showone=Imports#Imports) Do not import classes, functions, or other code elements of modules.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,72 @@
+
+
+def _sign_request(callback, token, error):
+  metadata = [('authorization', 'Bearer {}'.format(token))]
```
Never write a list literal (`[<expression>]`) where a tuple literal will do (`(<expression>,)`). Will a tuple literal do here?

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,72 @@
+    _sign_request(callback, self._access_token, None)
+
+
+def _sign_request(callback, token, error):
```
Please place this function earlier in the file than ("above") all of its use-sites.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/interop/methods.py:
```diff
@@ -39,12 +39,14 @@
 
 from oauth2client import client as oauth2client_client
 
+from grpc.beta import interfaces, implementations
```
One import per line, please.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_auth_test.py:
```diff
@@ -0,0 +1,95 @@
+# Copyright 2015, Google Inc.
```
If you authored this file this year, please date its header this year. If this file is a branch of an earlier file, should the code review tool be showing me something on the left side of this diff?

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_auth_test.py:
```diff
@@ -0,0 +1,95 @@
+    class MockPluginCallback:
+      def __call__(self, metadata, error):
+        test.assertEqual(metadata, [('authorization', 'Bearer token')])
+        test.assertEqual(error, None)
```
`assertIsNone` (or `assertIsNotNone`) when making comparisons with `None`.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_auth_test.py:
```diff
@@ -0,0 +1,95 @@
+        raise Exception()
+
+    call_creds = _auth.GoogleCallCredentials(MockGoogleCreds())
+    call_creds.__call__(None, MockPluginCallback())
```
`call_creds(None, MockPluginCallback())`?

kpayson64 commented:
I've addressed all the comments.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,73 @@
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""GRPCAuthMetadataPlugins for standard authentication."""
```
Blank line between license header and module doc string.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,73 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""GRPCAuthMetadataPlugins for standard authentication."""
+
+import concurrent.futures
```
`from concurrent import futures`.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,73 @@
+
+
+class GoogleCallCredentials(interfaces.GRPCAuthMetadataPlugin):
+  """Metadata wrapper for googleCredentials from the oauth2client library."""
```
Is the code element to which you're referring actually styled `googleCredentials`? Maintain consistent styling (case, underscores, and so forth) when referring to a code element in documentation.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/implementations.py:
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
+  """Construct CallCredentials from google credentials
+
+  Args:
+    credentials: A google credentials object from oauth2client
```
Don't be ambiguous about the type of this parameter - use the correct, formal name of the interface or class.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/implementations.py:
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
       _low.call_credentials_metadata_plugin(metadata_plugin, name))
 
+def google_call_credentials(credentials):
+  """Construct CallCredentials from google credentials
```
Missing period at the end.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/implementations.py:
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
+    credentials: A google credentials object from oauth2client
+
+  Returns:
+    A CallCredentials object for header authentication
```
Missing period at the end.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/implementations.py:
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
+  """Construct CallCredentials from a raw access token
+
+  Args:
+    access_token: A raw access token used for authentication
```
What's a "raw access token"? A byte string? An instance of some interface or class?

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_implementations_test.py:
```diff
@@ -48,6 +48,19 @@ def test_application_provided_root_certificates(self):
+class CallCredentialsTest(unittest.TestCase):
+
+  def test_google_call_credentials(self):
+    class MockGoogleCreds:
```
Missing inheritance from `object` - but why are you using a mock here anyway? Why not use a real instance?

[I have some opinions on the subject](https://www.youtube.com/watch?v=Xu5EhKVZdV8).

nathanielmanistaatgoogle commented on src/python/grpcio/tests/tests.json:
```diff
@@ -50,4 +53,4 @@
   "cygrpc_test.InsecureServerInsecureClient", 
   "cygrpc_test.SecureServerSecureClient", 
   "cygrpc_test.TypeSmokeTest"
-]
```
I suspect that you didn't intend the change on this line?

nathanielmanistaatgoogle commented on src/python/grpcio/tests/interop/client.py:
```diff
@@ -65,39 +65,35 @@ def _args():
         args.server_host_override)
     stub = test_pb2.beta_create_TestService_stub(
-        channel, metadata_transformer=metadata_transformer)
+        channel)
```
This line-wrap is now unnecessary, right?

kpayson64 commented:
I've addressed the comments.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/interop/client.py:
```diff
@@ -109,8 +104,7 @@ def _test_case_from_arg(test_case_arg):
   for test_case in methods.TestCase:
     if test_case_arg == test_case.value:
       return test_case
-  else:
```
Why this change? I prefer the `for`-`else`.

kpayson64 replied:
glint complained:
W0120:107:_test_case_from_arg: Else clause on loop without a break statement [useless-else-on-loop]

I've changed it back though.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_auth_test.py:
```diff
@@ -0,0 +1,95 @@
+
+      def get_access_token(self):
+        token = MockAccessToken()
+        token.access_token = 'token'
```
Concept/abstraction weirdness: why does something that is an access token have an `access_token` attribute? Access tokens aren't structurally recursive, are they?

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_auth_test.py:
```diff
@@ -0,0 +1,95 @@
+      self.assertIsNotNone(error)
+      callback_event.set()
+
+    class MockGoogleCreds(object):
```
This class doesn't require any of what's available to it in function-scope; please extract it to module-scope.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/implementations.py:
```diff
@@ -105,10 +106,37 @@ def metadata_call_credentials(metadata_plugin, name=None):
     A CallCredentials object for use in a GRPCCallOptions object.
   """
   if name is None:
-    name = metadata_plugin.__name__
```
I don't think I understand the change on this line? What's going on here?

kpayson64 replied:
By default python objects don't have a __name__ attribute.  If I pass in an arbitrary MetadataPlugin, it will throw an exception.

nathanielmanistaatgoogle replied:
It seems to me that an object's name, if present, is more likely to be helpful than its class name. Should this code look for the object name and only use the class name if it is not found? I believe @soltanmm originally authored this and may have an opinion.

kpayson64 replied:
@soltanmm 
As far as I can tell, objects having names is the exception and not the rule.  Is that the behavior we want, or are we just obfuscating naming unnecessarily?

soltanmm replied:
Woops, this one slipped by me. I think your change is the right way to go for non-function callables (which I think I'd not considered at the time). The problem is that we expect most users to just write a new function and pass it in. Functions have `__name__` attributes that are meaningful, but their `__class__` attributes are always `function`. So we'd at least want to maintain support for that use case.

soltanmm replied:
And, yes, that means we're supporting something that doesn't directly inherit from an interface we're requesting, and instead are duck typing, and haven't taken the time to do some ABC tricks to make it airtight, but it's straightforward enough that in offline discussion we determined it was desirable to implicitly support it.

kpayson64 replied:
I've updated the behavior to check for the function name first, and then fall back on the class name.

kpayson64 commented:
I've addressed the feedback.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_auth_test.py:
```diff
@@ -0,0 +1,98 @@
+    return token
+
+
+class MockExcepitonGoogleCreds(object):
```
Typo "Excepiton".

kpayson64 commented:
test this please

kpayson64 commented:
test this please

kpayson64 commented:
test this please

kpayson64 commented:
I've addressed the feedback, and also fixed a bug where MetadataPlugin destruction was causing a intermittent segfaults, which caused the interop tests to fail.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/beta/_auth.py:
```diff
@@ -0,0 +1,74 @@
+    future.add_done_callback(lambda x: self._get_token_callback(callback, x))
+
+  def _get_token_callback(self, callback, future):
+    access_token = None
```
No need for these fields at method top-level - use control flow instead:
```
try:
  access_token = future.result().access_token
except Exception as e:
  _sign_request(callback, None, e)
else:
  _sign_request(callback, access_token, None)
```
.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/beta/_auth_test.py:
```diff
@@ -0,0 +1,98 @@
+from grpc.beta import _auth
+
+
+class MockAccessTokenInfo(dict):
```
Why does this class need to exist? Why can't a `dict` be used instead at its one use-site?

nathanielmanistaatgoogle commented:
Looks good to me apart from two comments; @soltanmm please also review.

kpayson64 commented:
Made suggested changes and rebased.

nathanielmanistaatgoogle commented:
Luhfullyguhtum.

kpayson64 commented:
@soltanmm Friendly ping

soltanmm commented on src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi:
```diff
@@ -137,7 +137,7 @@ cdef void plugin_get_metadata(
   self.plugin_callback(cy_context, python_callback)
 
-cdef void plugin_destroy_c_plugin_state(void *state):
+cdef void plugin_destroy_c_plugin_state(void *state) with gil:
```
Sorry for the long round trip. What's up with the added GIL acquisition? Does the decref go boom without it? I could've sworn Cython was supposed to whine if acquiring the GIL was needed to avoid going boom.

kpayson64 replied:
I'm not sure about Cython's behavior, but it seems like manipulating python objects in c extensions requires the gil

https://docs.python.org/2/c-api/init.html#thread-state-and-the-global-interpreter-lock

The error I was seeing was a consistent segfault on the first execution of the interop client in a fresh docker instance, and this change fixed it.

soltanmm replied:
Huh. Arguably Cython's supposed to statically error out if such a thing is necessary and not present, so that's something to file on their tracker at some point, methinks...

But otherwise, yeah, works for me.

soltanmm replied:
Actually, this may not be one for the Cython tracker. May need to investigate a little.

soltanmm replied:
Ah. The problem is that Cython correctly sees that, from Python code, this function is only ever called when the GIL is held (due to its old signature implicitly requesting that calling code hold the GIL, else it'd be `nogil`), but we let this function's pointer escape into a non-Python non-Cython function, so the transpiler can't statically enforce that guarantee.

soltanmm commented:
lgtm on tests

kpayson64 commented:
test this please

kpayson64 commented:
All relevant tests have passed.
